### PR TITLE
power standardization 1/??????? - power scales, SMES/cellrate/APC/powernet/god i'm dying inside

### PR DESCRIPTION
**Pull request by silicons**
what was supposed to be a 30 minute refactor has been 2 days and counting because it turns out 16 MW is the biggest you can get with a scale of 1 rather than 1000 on 32 bit floats so

- fuck byond
- fuck this
- fuck SMES
- fuck power cells
- this is still more fun than responding to my discord DM queue

### CODE REVIEW

**Comment by moonheart08** [comment]
i'm required to inform you i strongly disapprove of putting the between -> clamp change in this PR, but I cannot stop you.

File: code/game/gamemodes/technomancer/spells/energy_siphon.dm [hunk 1]
```diff
@@ -102,14 +100,13 @@
...
-			if(AM.drain_power(1) <= 0)
+			if(!AM.can_drain_energy(src, NONE))
 				things_to_drain.Remove(AM)
 		if(!things_to_drain.len)
 			return
-		var/charge_to_steal = round(flow_remaining / things_to_drain.len) // This is to drain all the cells evenly.
 		for(var/atom/movable/AM in things_to_drain)
-			var/big_number = AM.drain_power(0,0,charge_to_steal / CELLRATE) // This drains the cell, and leaves us with a big number.
-			flow_remaining = flow_remaining - (big_number * CELLRATE) // Which we reduce to our needed number by multiplying.
+			var/big_number = AM.drain_energy(user, flow_rate * 0.001) * 1000 // This drains the cell, and leaves us with a big number.
```
Comment (moonheart08) [inline code/game/gamemodes/technomancer/spells/energy_siphon.dm:108]: Magic numbers? Anything stopping you from at least using a define here?

File: code/game/machinery/cell_charger.dm [hunk 2]
```diff
@@ -7,17 +7,16 @@
 	use_power = USE_POWER_IDLE
 	power_channel = EQUIP
 	idle_power_usage = 5
-	active_power_usage = 60000 //60 kW. (this the power drawn when charging)
+	active_power_usage = 30000 //60 kW. (this the power drawn when charging)
```
Comment (moonheart08) [inline code/game/machinery/cell_charger.dm:10]: pre-multiplied by magic constant? See above.

File: code/game/machinery/cell_charger.dm [hunk 3]
```diff
@@ -132,7 +130,8 @@
 	var/E = 0
 	for(var/obj/item/stock_parts/capacitor/C in component_parts)
 		E += C.rating
-	efficiency = active_power_usage * (1+(E-1)*0.5) * 10
+	update_active_power_usage(30000 * E)
```
Comment (moonheart08) [inline code/game/machinery/cell_charger.dm:133]: inlined a mutable variable, is this intended?

File: code/game/machinery/exonet_node.dm [hunk 4]
```diff
@@ -25,7 +25,6 @@
 	. = ..()
 
 	default_apply_parts()
-	RefreshParts()
```
Comment (moonheart08) [inline code/game/machinery/exonet_node.dm:28]: Intended removal? seems unrelated.

File: code/game/machinery/spaceheater.dm [hunk 5]
```diff
@@ -324,7 +324,7 @@
 /obj/machinery/power/thermoregulator/overload(var/obj/machinery/power/source)
 	if(!anchored || !powernet)
 		return
-	var/power_avail = draw_power(active_power_usage*10)
+	var/power_avail = draw_power(active_power_usage * 10 * 0.001) * 1000
```
Comment (moonheart08) [inline code/game/machinery/spaceheater.dm:327]:
```suggestion
	var/power_avail = draw_power(active_power_usage * 0.01) * 100
```
for if you really insist on the magic constants, here's one less multiply.

File: code/game/mecha/mecha.dm [hunk 6]
```diff
@@ -214,15 +214,13 @@
 			var/obj/item/mecha_parts/mecha_equipment/ME = new path(src)
 			ME.attach(src)
 
-/obj/mecha/drain_power(var/drain_check)
-
-	if(drain_check)
-		return 1
-
+/obj/mecha/drain_energy(datum/acter, amount, flags)
```
Comment (moonheart08) [inline code/game/mecha/mecha.dm:217]: Is acter a typo?

File: code/modules/atmospherics/components/binary_devices/massive_gas_pump.dm [hunk 7]
```diff
@@ -86,7 +86,7 @@
 	if (power_draw >= 0)
 		last_power_draw = power_draw
 
-		power_machine.draw_power(power_draw)
+		power_machine.draw_power(power_draw * 0.001)
```
Comment (moonheart08) [inline code/modules/atmospherics/components/binary_devices/massive_gas_pump.dm:89]: Seemingly redundant multiplication, you're just multiplying a value by 1000 only to divide it back later. I'm assuming pump_gas doesn't care what unit it's being fed.

File: code/modules/atmospherics/components/binary_devices/massive_heat_pump.dm [hunk 8]
```diff
@@ -95,7 +95,7 @@
 	if (power_draw >= 0)
 		last_power_draw = power_draw
 
-		power_machine.draw_power(power_draw)
+		power_machine.draw_power(power_draw * 0.001)
```
Comment (moonheart08) [inline code/modules/atmospherics/components/binary_devices/massive_heat_pump.dm:98]: Redundant, precision eating multiply. See above.

File: code/modules/mob/living/simple_mob/subtypes/vore/solargrub.dm [hunk 9]
```diff
@@ -96,21 +98,20 @@ GLOBAL_LIST_EMPTY(solargrubs)
...
+			charge += power_drain
 			for(var/obj/machinery/power/terminal/T in PN.nodes)
 				if(istype(T.master, /obj/machinery/power/apc))
 					var/obj/machinery/power/apc/A = T.master
 					if(A.operating && A.cell)
-						var/cur_charge = A.cell.charge / CELLRATE
-						var/drain_val = min(apc_drain_rate, cur_charge)
-						A.cell.use(drain_val * CELLRATE)
+						// they're now a threat
+						// but also fuck off with your *pawstomp* comments ~silicons
```
Comment (moonheart08) [inline code/modules/mob/living/simple_mob/subtypes/vore/solargrub.dm:108]: it was "*pawstamp*"
at least get it right :heck:

File: code/modules/power/apc.dm [hunk 10]
```diff
@@ -146,30 +152,24 @@ GLOBAL_LIST_EMPTY(apcs)
...
-	if(terminal && terminal.powernet)
+	if(terminal?.powernet)
 		terminal.powernet.trigger_warning()
-		drained_energy += terminal.powernet.draw_power(amount)
+		// no conversion - amount = kj, draw_power is in kw
+		drained += terminal.powernet.draw_power(amount)
 
 	//The grid rarely gives the full amount requested, or perhaps the grid
 	//isn't connected (wire cut), in either case we draw what we didn't get
 	//from the cell instead.
```
Comment (moonheart08) [inline code/modules/power/apc.dm:168]: Maybe double-check that your subtraction of power from the grid actually did anything, if possible. If it didn't, draw only from the cell. (if the cell is below, i'unno, half full, pretend it did and refill it anyways. At that point precision is too far gone to be saveable.)

This'll allow you to avoid APCs using 0 power when precision gets too out of wack. However considering you redid the scaling it's likely unnecessary, in which case disregard.

File: code/modules/power/fusion/magpower.dm [hunk 11]
```diff
@@ -1,4 +1,4 @@
-#define ENERGY_PER_K 20
+#define ENERGY_KW_PER_K 20
```
Comment (moonheart08) [inline code/modules/power/fusion/magpower.dm:1]: Forgot a unit conversion?

**Review by moonheart08 — COMMENTED** [review]
As is this doesn't seem to solve the fundamental issue that large number X + 0.05 == X

Best way to approach it would be to accumulate all power to be added and removed for the tick, so that summing all power changes together is less likely to entirely lose value.

Another approach is to force APCs and power providers to always round to exact KW, and store any excess drainage until it adds up to a KW. This avoids 50-100W differences being lost to precision past 524288KW, and allows you to never lose any power to precision when total draw is under 2^24KW.

Essentially:
```dm
draw_power(floor(power_usage) + floor(accumulator))
if (floor(accumulator) >= 1)
   accumulator = accumulator - floor(accumulator) // remove power that got drawn by the powernet.
accumulator += power_usage - floor(power_usage) // add power that didn't get drawn immediately to the accumulator.
```
I didn't do a full review, I can finish it if you'd like.

**Comment by silicons** [comment]
OK SO
I am going to unfuck draw_power and make APCs accumulator
ANYTHIGN ELSE IS GOING IN ANOTHER PR GOD I AM DYING INSIDE I HATE BYOND I HATE BYOND I HATEB YOND AAAAAAAAAA
